File: 84.largest-rectangle-in-histogram.js
/*
 * @lc app=leetcode id=84 lang=javascript
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
/**
 * @param {number[]} heights
 * @return {number}
 */
const largestRectangleArea = function (heights)
{
    heights.push(0);

    let maxArea = 0;

    /**
     * 存储高度下标，按照高度递增
     * @type {number[]}
     */
    const monostack = [];
    for (let i = 0; i < heights.length; i++)
    {
        if (monostack.length === 0)
        {
            monostack.push(i);
        }
        else
        {
            const topIndex = monostack[monostack.length - 1];
            const topHeight = heights[topIndex];
            if (topHeight < heights[i])
            {
                monostack.push(i);
            }
            else    // topHeight >= heights[i]
            {
                monostack.pop();
                const poppedTopIndex = monostack[monostack.length - 1];

                maxArea = Math.max(maxArea, topHeight * (
                    monostack.length === 0
                        ? i
                        : (i - poppedTopIndex - 1)));
                i--;
            }
        }
    }


    return maxArea;
};
// @lc code=end

File: Daily/84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */
#include <algorithm>
#include <stack>
#include <vector>

using std::stack;
using std::vector;

// @lc code=start
class Solution {
 public:
  int largestRectangleArea(vector<int>& heights) {
    heights.push_back(0); // force stack to empty in the end

    int maxArea = 0;
    stack<int> increasingIndexStack;
    for (int i = 0; i < heights.size(); i++) {
      while (!increasingIndexStack.empty() &&
             heights[increasingIndexStack.top()] >= heights[i]) {
        int topIndex = increasingIndexStack.top();
        increasingIndexStack.pop();

        int topHeight = heights[topIndex];
        int topHeightWidth = increasingIndexStack.empty()
                           ? i
                           : i - increasingIndexStack.top() -
                                 1;
        maxArea = std::max(maxArea, topHeight * topHeightWidth);
      }
      increasingIndexStack.push(i);
    }

    return maxArea;
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<int> vec = {1, 2, 2};  // 4
  sol.largestRectangleArea(vec);
}